### ISSUE TITLE
Probably inappropriate proposals for function argument lintr

### ISSUE THREAD

**Issue by strengejacke** [r-lib/lintr#2598]
We had a case where we deprecated argument names. The default value of the new (and formerly, of the old) argument is `"all"`. We set the old argument to `NULL` and used `is.null()` to check whether users provided a value for the old argument. However, since `NULL` as also a valid value, using `old_argument = NULL` intentionally - in combination with checking for `is.null()`, resulted in ignoring this and using the new argument (with its _different_ default value, `"all"`). This actually caused severe memory allocation problems, even crashing test machines on CRAN.

We than fixed this and did not provide a default value for the old argument, and checked if users provided a value using `missing(old_argument)`. After we did this change, following lintr message occured:

> Warning: file=,[function_argument_linter] Arguments without defaults should come before arguments with defaults. Consider setting the default to `NULL` and using `is.null()` instead of using `missing()`

While it's ok to warn about arguments w/o defaults should be placed before arguments _with_ defaults, the suggestion can be very problematic in edge cases as ours. lintr suggests what we have just "fixed" and what can lead to critical issues. Maybe either change the suggestion completely, or add

"Consider setting the default to `NULL` and using `is.null()` instead of using `missing()` **if the former default value was NULL as well**" (or something similar)?

**Comment by MichaelChirico** [comment]
Sorry, could you provide more specific context? what was the signature before and after?

My instinct is you should just `#nolint` for this case. this performance issue also sounds like some sort of bug, not sure where best to report it.

**Comment by strengejacke** [comment]
We have a function that creates data grids (for predictions), `insight::get_datagrid()`. The `at` argument specified at which values the grid should be constructed. Default value was `"all"`, which creates larger grids. If you set `at = NULL`, a small grid is returned.

Now we renamed `at` into `by`. `by = "all"` by default (like the former `at`), but we set `at = NULL` and checked inside the code whether `!is.null(at)`. If `at` was `NULL`, it was ignored. However, setting `at = NULL` was a valid option (to return a small grid), and thus, using `insight::get_datagrid(at = NULL)` was equivalent to `insight::get_datagrid(by = "all")` (since `at` was ignored if `NULL`). This unintentionally lead to very large data grids in edge cases.

Then we fixed this and had `at` w/o value as function argument. Inside `insight::get_datagrid()`, we changed `!is.null(at)` into `missing(at)`. When the user now still uses the old argument (`insight::get_datagrid(at = NULL)`), we can through a deprecation warning and set `by = NULL` (as intended by the user), to return the desired result.

But: Now lintr warns "Arguments without defaults should come before arguments with defaults" (because `at` is at the end of arguments in that function definition, and w/o value), which is ok. However, the *suggestion* "Consider setting the default to NULL and using is.null() instead of using missing()" is dangerous for the above described reasons. This is an edge case, but probably hard to debug if it occurs.

Thus, my suggestion would be to either remove "Consider setting the default to NULL and using is.null() instead of using missing()", or to revise it, e.g. "Consider keeping the old default value and check for it, instead of using missing()"

**Comment by AshesITR** [comment]
I'm not sure we can help with this type of regression.
As said, for your issue, a `# nolint: function_argument.` would be the recommended way to go.

BTW: You could use an S3 based approach with a sentinel value `deprecated()` instead of removing the default.

**Comment by strengejacke** [comment]
I'm not complaining about the lintr warning itself. It's the recommendation the package gives to resolve lintr issues.

TL;DR:

> Warning: file=,[function_argument_linter] Arguments without defaults should come before arguments with defaults. ***Consider setting the default to NULL and using is.null() instead of using missing()***

The highlighted recommendation is dangerous for edge cases when `NULL` is a valid option (in which case it might be better to use `missing()`).

**Comment by strengejacke** [comment]
I haven't followed the recommendation, because I **had** used `NULL` before (leading to crash R), then removed the default for the argument and switched to use `missing()`. **Then** I got the lintr message, recommending what I just had fixed (because, as I said, in my situation, it caused even CRAN machines to crash checks).

**Comment by strengejacke** [comment]
Here's a reprex:

``` r
# old function, with old argument "at"
foo <- function(data, at = "all") {
  if (is.null(at)) {
    expand.grid(lapply(iris, unique))
  } else if (at == "all") {
    expand.grid(data)
  }
}

nrow(foo(iris)) # at = "all" creates a too large grid
#> Error: cannot allocate vector of size 282.9 Gb

nrow(foo(iris, at = NULL)) # NULL is a valid option
#> [1] 2284590


# at is deprecated, use by instead. We set "at = NULL" now,
# so we can test with "is.null()" whether user used the
# deprecated argument. Now using at = NULL as valid option
# no longer works, because it's ignored.
foo_new <- function(data, by = "all", at = NULL) {
  if (!is.null(at)) {
    warning("at is deprecated, use by instead", call. = FALSE)
    by <- at
  }
  if (is.null(by)) {
    expand.grid(lapply(iris, unique))
  } else if (by == "all") {
    expand.grid(data)
  }
}

nrow(foo_new(iris))
#> Error: cannot allocate vector of size 282.9 Gb

nrow(foo_new(iris, at = NULL)) # NULL is a valid option
#> Error: cannot allocate vector of size 282.9 Gb


# we need to fix this, as we cannot use at = NULL and "is.null()".
# we now remove the default value for the at argument and check with
# "missing()" instead, because as seen in "foo_new()", at = NULL as valid
# option doesn't work and causes serious trouble. 
# lintr complaints, but thats ok
foo_new2 <- function(data, by = "all", at) { 
  # but... lintr recommend what we have done above in "foo_new", which
  # doesn't work! lintr warning is ok -  but bad recommendation how to fix!
  if (!missing(at)) {
    warning("at is deprecated, use by instead", call. = FALSE)
    by <- at
  }
  if (is.null(by)) {
    expand.grid(lapply(iris, unique))
  } else if (by == "all") {
    expand.grid(data)
  }
}

nrow(foo_new2(iris))
#> Error: cannot allocate vector of size 282.9 Gb

nrow(foo_new2(iris, at = NULL)) # NULL is a valid option, works again
#> Warning: at is deprecated, use by instead
#> [1] 2284590
```

<sup>Created on 2024-06-11 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by AshesITR** [comment]
I understand. Do you have a suggestion that doesn't assume that the lint is the result of refactoring a previously existing argument?
I'm afraid that would cause confusion whenever the lint occurs in a context other than argument deprecation, such as when writing a new function or when adding a default to some middle argument of an existing function without defaults.

Illustrative example:

```r
foo_old <- function(x, y, z) {
  42L
}

foo_new <- function(x, y = 2L, z) {
  y
}
```

**Comment by MichaelChirico** [comment]
Your example is why the message is worded as "consider..." -- it doesn't mean it applies in every case and certainly yours is one where it doesn't apply.

As noted your suggested rewording goes too far in assuming other users are in the same situation. Do you have a different suggested wording that would have made the "consider..." part less subtle?

**Comment by IndrajeetPatil** [comment]
I think the current phrasing already achieves the delicate balance of not overgeneralizing and not being tied to too specific a case. 

So nothing more to do here.

**Comment by IndrajeetPatil** [comment]
I think the current phrasing already achieves the delicate balance of not overgeneralizing and not being tied to too specific a case. 

So nothing more to do here.
